## Building a recommendation engine to recommend books in Spark

Using collaborative filtering to predict ratings of unread books on 'Goodreads'

Warren Buffett was once asked about the key to success, he pointed to a stack of nearby books and said, "Read 500 pages like this every day. That's how knowledge works. It builds up, like compound interest. All of you can do it, but I guarantee not many of you will do it."
Books are the best resources for most of us to develop and gain perspectives. 

I myself love reading books. Once I like a book, I have a habit of going to good books or asking someone who has a similar taste to look for recommendations for the next series of books I might like. 
Artificial Intelligence has made our world so easy by recommending us books, movies, and products all based on the past data that saves our time and energy on analyzing different options. In fact, sometimes machines can recommend us better than what we think because they don't suffer from emotional biases.

**The intuition behind Alternating Least Square**

For someone who loves reading academic papers, here's the link to the paper: https://datajobs.com/data-science-repo/Recommender-Systems-%5BNetflix%5D.pdf.

The most important kind of recommender system is collaborative filtering based approach. Let's say you know a friend who has the same taste as you because you both love psychology, then you might like reading other books that your friend has read but you haven't. This is the sole concept behind collaborative filtering. Now let's talk specifically about Alternating least square method.


Collaborative filtering can be easily achieved by matrix factorization techniques like Singular Value decomposition where a user-rating matrix is decomposed into the user-concept matrix, concept-weights matrix, and rating-concept matrix. Concepts are basically latent or hidden factors that the matrix decomposition implicitly generates like in the case of the books, the different concepts can be psychology, data science, philosophy, etc.

Most of the matrix factorization techniques like Singular Value decomposition don't know how to deal with an incomplete/sparse matrix which means having empty values in the user-rating matrix(which is common as not every user would have read all the books). Traditionally, engineers have been imputing those values with the mean or median before performing matrix factorization to do collaborative filtering. This leads to overfitting since the books that have never been rated are being imputed by the mean or median which can skew the results towards them.

Recent methods like Alternating Least square don't suffer from these fallbacks. They suggest modeling directly the observed ratings while avoiding overfitting through a regularized mode

In [2]:
#import to_read data

file_location = "/FileStore/tables/to_read.csv"
file_type = "csv"

# CSV options
infer_schema = "True"
first_row_is_header = "True"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
to_read = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(to_read)

# Create Temporary Tables
to_read.createOrReplaceTempView("to_read")

user_id book_id 1 112 1 235 1 533 1 1198 1 1874 1 2058 1 3334 2 4 2 11 2 13 2 16 2 92 2 106 2 113 2 233 2 247 2 380 2 408 2 439 2 479 2 516 2 619 2 1069 2 1371 2 1384 2 1597 2 1926 2 1967 2 1969 2 2220 2 2388 2 3222 2 3227 2 3733 2 5418 2 6937 2 7146 2 7873 2 8417 3 10 3 66 3 233 3 726 3 932 3 1146 3 2439 3 3566 3 4582 3 5448 3 5765 3 6274 3 8403 5 31 5 665 5 3074 5 3581 6 64 6 459 6 619 6 649 6 687 6 874 6 1319 6 1912 6 5269 6 5537 6 5820 6 6220 6 7480 7 10 7 215 7 225 7 233 7 1322 7 3552 7 3720 7 4468 8 19 8 498 8 1357 8 2177 8 3930 8 4004 8 8792 8 9304 9 8 9 112 9 139 9 230 9 233 9 473 9 638 9 653 9 672 9 701 9 736 9 883 9 895 9 1467 9 1582 9 1873 9 2516 9 2630 9 3210 9 3476 9 3560 9 3650 9 5286 9 6636 10 6 10 385 10 662 10 896 10 1076 10 1082 10 1220 10 1281 10 1439 10 1851 10 1920 10 2441 10 2570 10 2616 10 3059 10 3545 10 3961 10 4369 10 4510 10 5104 10 5254 10 5556 10 6114 10 6917 10 7545 10 7842 11 3643 12 1 12 124 12 1471 12 2893 13 81 13 277 13 318 13 430 13 437 13 551 13 560 13 610 13 678 13 979 13 1082 13 1464 13 1577 13 2018 13 2418 13 3257 13 4212 13 4635 13 4704 13 4734 13 4841 13 4848 13 4856 13 4928 13 6114 13 6793 13 6962 13 8717 14 493 14 1179 14 2693 14 3581 14 3683 14 3953 14 4220 14 5076 14 5537 14 7777 15 191 15 275 15 398 15 666 15 1491 15 5210 17 1344 20 7628 22 653 22 1403 22 1405 22 1645 22 2500 22 2691 22 4995 23 273 23 459 23 791 23 886 23 925 23 1198 23 1597 23 1688 23 2251 23 2350 23 2613 23 3312 23 3343 24 61 24 143 24 421 24 467 24 974 26 508 26 1796 26 2257 26 2314 26 2357 26 2738 26 2810 26 3207 26 3412 26 3586 26 5091 26 5550 26 6877 26 7970 28 265 28 587 28 3543 28 5215 28 7496 28 8654 29 31 29 32 29 36 29 45 29 65 29 87 29 115 29 121 29 131 29 178 29 256 29 323 29 424 29 467 29 550 29 618 29 662 29 709 29 730 29 771 29 809 29 934 29 1084 29 1364 29 1642 29 1695 29 2259 29 2284 29 2304 29 2967 29 3508 29 4359 29 4394 29 4475 29 4629 29 5090 29 5596 29 5851 29 6104 30 47 30 67 30 82 30 121 30 177 30 249 30 297 30 407 30 426 30 430 30 442 30 516 30 658 30 895 30 1039 30 1094 30 1381 30 2117 30 2426 30 3444 30 4004 30 4202 30 4509 30 5880 30 5907 30 6828 30 7563 30 8737 31 94 31 584 31 1682 31 3417 32 11 32 76 32 118 32 121 32 133 32 261 32 272 32 1509 32 1923 32 2897 32 3646 33 45 33 146 33 185 33 208 33 352 33 610 33 677 33 1052 33 1527 33 1708 33 3300 33 5776 33 7071 34 113 34 215 34 312 34 380 34 483 34 593 34 850 34 1574 34 1591 34 1629 34 3109 34 3581 34 7414 34 8598 35 139 35 198 36 7 36 55 36 65 36 612 36 624 36 631 36 652 36 718 36 724 36 741 36 748 36 1077 36 2415 36 6307 37 346 37 819 37 7173 39 17 39 20 39 58 39 59 39 93 39 117 39 125 39 154 39 188 39 219 39 229 39 298 39 353 39 386 39 400 39 408 39 574 39 595 39 614 39 646 39 665 39 714 39 773 39 809 39 819 39 899 39 914 39 959 39 1081 39 1167 39 1283 39 1367 39 1420 39 1486 39 1566 39 1570 39 1797 39 2240 39 2485 39 2543 39 2601 39 2727 39 2926 39 3020 39 3022 39 4196 39 4359 39 4973 39 5327 39 5578 39 7569 39 8404 39 9114 39 9946 40 385 40 688 40 1657 41 205 41 727 41 1766 41 5574 41 7486 43 7612 44 4111 45 342 46 87 46 105 46 119 46 124 46 135 46 206 46 535 46 584 46 638 46 771 46 822 46 1105 46 1439 46 1761 46 2288 46 2496 46 2543 46 3472 46 8479 47 45 47 137 47 551 47 1281 47 1708 47 2583 47 4216 47 4282 47 4307 47 5416 47 7193 47 8329 47 8957 49 754 50 323 50 430 50 941 50 1060 50 1421 50 3187 50 3296 51 141 51 816 51 949 51 1187 51 1415 51 2239 51 2947 51 3024 51 3695 51 5378 51 5425 51 5779 51 6650 54 146 54 190 54 193 54 267 54 333 54 654 54 749 54 867 54 1198 54 1874 54 2368 54 2498 54 2627 54 3118 54 3194 54 3236 54 3812 54 5794 54 7770 54 8230 54 8693 55 127 55 291 55 930 55 2374 55 3021 56 239 56 329 56 385 56 551 56 813 56 836 56 896 56 1009 56 1174 56 1195 56 1220 56 1242 56 1739 56 1920 56 2088 56 2334 56 2449 56 2543 56 2775 56 2787 56 2840 56 3253 56 3946 56 3955 56 4116 56 4635 56 4967 56 5418 56 5776 56 6307 56 7397 56 8498 56 8867 56 9054 57 352 57 398 57 508 57 533 57 610 57 983 57 1017 57 1061 57 1220 57 174

In [3]:
#import ratings data
file_location = "/FileStore/tables/ratings.csv"
file_type = "csv"

# CSV options
infer_schema = "True"
first_row_is_header = "True"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
ratings = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(ratings)

# Create Temporary Tables
ratings.createOrReplaceTempView("ratings")

book_id user_id rating 1 314 5 1 439 3 1 588 5 1 1169 4 1 1185 4 1 2077 4 1 2487 4 1 2900 5 1 3662 4 1 3922 5 1 5379 5 1 5461 3 1 5885 5 1 6630 5 1 7563 3 1 9246 1 1 10140 4 1 10146 5 1 10246 4 1 10335 4 1 10610 5 1 10944 5 1 11854 4 1 11927 4 1 12471 5 1 13282 5 1 13544 5 1 15494 5 1 16377 4 1 16913 5 1 17434 5 1 17663 5 1 17984 5 1 18031 5 1 18313 5 1 18361 4 1 20076 3 1 20467 3 1 20848 5 1 21228 5 1 21487 4 1 21713 5 1 22602 4 1 23576 4 1 23612 4 1 24326 5 1 24389 4 1 24499 5 1 24834 4 1 24845 5 1 25164 4 1 25182 5 1 25214 4 1 26145 4 1 26629 4 1 26661 4 1 28158 4 1 28767 5 1 29123 3 1 29703 5 1 30681 5 1 31001 4 1 32055 4 1 32305 5 1 32592 4 1 32635 5 1 32748 5 1 32923 5 1 33065 4 1 33697 4 1 33716 5 1 33872 5 1 33890 3 1 37284 5 1 37834 5 1 38080 4 1 38082 5 1 38475 4 1 39423 3 1 41074 3 1 42404 5 1 43985 4 1 44243 2 1 44397 5 1 45269 4 1 45493 5 1 46977 4 1 47476 4 1 47746 5 1 47800 5 1 48482 3 1 49298 4 1 50104 5 1 50342 3 1 51166 4 1 51460 3 1 51480 1 1 51838 5 1 52036 2 1 53245 4 2 1169 3 2 3022 4 2 3922 5 2 5115 4 2 5436 5 2 5461 4 2 5885 4 2 6063 1 2 6342 3 2 6630 5 2 8167 2 2 9731 4 2 10111 5 2 10140 3 2 10288 5 2 10509 2 2 10610 5 2 10751 3 2 10944 5 2 11285 4 2 11408 5 2 11691 4 2 11692 3 2 11868 5 2 11927 5 2 11945 5 2 12874 4 2 12946 5 2 13282 5 2 13794 1 2 14372 3 2 14546 5 2 14603 4 2 15604 4 2 16913 2 2 17434 5 2 17566 4 2 17643 1 2 17663 5 2 17984 5 2 18313 5 2 18361 4 2 19526 4 2 19724 5 2 19729 4 2 19942 5 2 21217 4 2 21228 4 2 21487 5 2 21676 5 2 21733 5 2 22602 4 2 23576 5 2 23612 5 2 24499 5 2 24834 4 2 25164 4 2 25182 5 2 26145 4 2 26629 4 2 27499 5 2 28767 4 2 30313 5 2 30681 4 2 30944 5 2 32055 4 2 32305 5 2 32592 4 2 32635 4 2 32748 5 2 32918 5 2 33065 5 2 33697 4 2 33716 5 2 33872 5 2 36099 5 2 37834 4 2 38082 5 2 42404 5 2 42508 4 2 42810 4 2 45269 4 2 45493 5 2 46421 5 2 46977 3 2 47476 3 2 47478 2 2 47800 5 2 48482 5 2 48559 5 2 48687 1 2 49298 5 2 50096 5 2 50104 5 2 50133 5 2 51166 5 2 51460 4 2 51838 4 2 52036 3 2 53292 5 3 314 3 3 588 1 3 2077 2 3 2487 3 3 2900 3 3 3662 4 3 4536 1 3 5115 4 3 5272 5 3 5461 2 3 5885 4 3 6063 5 3 6630 5 3 6634 3 3 7001 4 3 9246 1 3 9731 2 3 9771 3 3 10140 3 3 10246 1 3 10249 5 3 10509 1 3 10610 1 3 10751 1 3 10944 1 3 11285 2 3 11408 3 3 11599 4 3 11854 1 3 11927 4 3 12471 2 3 12946 4 3 13274 5 3 13282 5 3 13794 1 3 14546 5 3 15604 1 3 16377 1 3 16569 1 3 16913 4 3 17434 4 3 17663 4 3 18031 3 3 18313 5 3 19724 4 3 19942 4 3 19984 2 3 20076 3 3 20782 4 3 20848 4 3 21228 3 3 21487 3 3 21733 1 3 22602 3 3 23576 1 3 23612 3 3 24326 3 3 24389 3 3 25182 3 3 25214 1 3 26398 2 3 26629 4 3 28767 3 3 29703 1 3 30681 4 3 32305 1 3 32592 5 3 32635 4 3 32918 5 3 33065 1 3 33716 3 3 36099 4 3 36240 2 3 37284 1 3 37834 4 3 37855 3 3 38080 5 3 38082 3 3 38475 3 3 42404 5 3 43602 4 3 43985 2 3 46977 2 3 47746 5 3 48482 3 3 48687 1 3 49288 5 3 49289 5 3 49295 5 3 49297 5 3 49298 1 3 50096 5 3 50101 5 3 50102 5 3 50104 5 3 50133 5 3 51166 3 3 51460 2 3 52036 1 3 53292 5 4 439 5 4 1169 5 4 2324 5 4 2487 5 4 3739 5 4 3922 4 4 4606 1 4 5115 5 4 5436 5 4 6342 5 4 6630 4 4 8167 5 4 8178 5 4 9246 4 4 10140 4 4 11285 3 4 11691 5 4 11692 5 4 11868 3 4 11945 5 4 12381 5 4 12874 5 4 12946 5 4 13034 4 4 13776 4 4 13794 4 4 14192 4 4 14207 3 4 14546 5 4 14603 5 4 15604 3 4 17228 5 4 17643 5 4 17663 5 4 17984 5 4 18031 5 4 18313 5 4 19526 5 4 19724 5 4 19729 4 4 20406 2 4 20467 4 4 21217 4 4 21228 3 4 21487 5 4 21676 3 4 21733 5 4 23576 5 4 24326 5 4 25164 5 4 25182 5 4 25840 3 4 26579 3 4 26629 4 4 26661 5 4 27499 5 4 27834 5 4 27934 4 4 28767 5 4 29123 3 4 29703 5 4 30283 4 4 30313 5 4 30681 5 4 30944 5 4 32305 5 4 32338 4 4 32592 5 4 32745 4 4 32748 5 4 32918 5 4 32923 5 4 33065 5 4 33697 5 4 33716 4 4 33872 5 4 34688 4 4 36099 5 4 37284 3 4 37834 4 4 39423 5 4 42404 5 4 43985 5 4 44243 3 4 44397 5 4 45269 5 4 46482 5 4 47411 4 4 47730 5 4 47800 5 4 48482 3 4 48559 5 4 48687 5 4 49298 5 4 50102 5 4 51460 5 4 51838 4 4 52036 3 4 53245 5 4 53292 5 5 314 4 5 1169 5 5 1952 5 5 2324 5 5 2487 3

Let's see the distribution of **average ratings**

In [5]:
%sql

select * from ratings

book_id user_id rating 1 314 5 1 439 3 1 588 5 1 1169 4 1 1185 4 1 2077 4 1 2487 4 1 2900 5 1 3662 4 1 3922 5 1 5379 5 1 5461 3 1 5885 5 1 6630 5 1 7563 3 1 9246 1 1 10140 4 1 10146 5 1 10246 4 1 10335 4 1 10610 5 1 10944 5 1 11854 4 1 11927 4 1 12471 5 1 13282 5 1 13544 5 1 15494 5 1 16377 4 1 16913 5 1 17434 5 1 17663 5 1 17984 5 1 18031 5 1 18313 5 1 18361 4 1 20076 3 1 20467 3 1 20848 5 1 21228 5 1 21487 4 1 21713 5 1 22602 4 1 23576 4 1 23612 4 1 24326 5 1 24389 4 1 24499 5 1 24834 4 1 24845 5 1 25164 4 1 25182 5 1 25214 4 1 26145 4 1 26629 4 1 26661 4 1 28158 4 1 28767 5 1 29123 3 1 29703 5 1 30681 5 1 31001 4 1 32055 4 1 32305 5 1 32592 4 1 32635 5 1 32748 5 1 32923 5 1 33065 4 1 33697 4 1 33716 5 1 33872 5 1 33890 3 1 37284 5 1 37834 5 1 38080 4 1 38082 5 1 38475 4 1 39423 3 1 41074 3 1 42404 5 1 43985 4 1 44243 2 1 44397 5 1 45269 4 1 45493 5 1 46977 4 1 47476 4 1 47746 5 1 47800 5 1 48482 3 1 49298 4 1 50104 5 1 50342 3 1 51166 4 1 51460 3 1 51480 1 1 51838 5 1 52036 2 1 53245 4 2 1169 3 2 3022 4 2 3922 5 2 5115 4 2 5436 5 2 5461 4 2 5885 4 2 6063 1 2 6342 3 2 6630 5 2 8167 2 2 9731 4 2 10111 5 2 10140 3 2 10288 5 2 10509 2 2 10610 5 2 10751 3 2 10944 5 2 11285 4 2 11408 5 2 11691 4 2 11692 3 2 11868 5 2 11927 5 2 11945 5 2 12874 4 2 12946 5 2 13282 5 2 13794 1 2 14372 3 2 14546 5 2 14603 4 2 15604 4 2 16913 2 2 17434 5 2 17566 4 2 17643 1 2 17663 5 2 17984 5 2 18313 5 2 18361 4 2 19526 4 2 19724 5 2 19729 4 2 19942 5 2 21217 4 2 21228 4 2 21487 5 2 21676 5 2 21733 5 2 22602 4 2 23576 5 2 23612 5 2 24499 5 2 24834 4 2 25164 4 2 25182 5 2 26145 4 2 26629 4 2 27499 5 2 28767 4 2 30313 5 2 30681 4 2 30944 5 2 32055 4 2 32305 5 2 32592 4 2 32635 4 2 32748 5 2 32918 5 2 33065 5 2 33697 4 2 33716 5 2 33872 5 2 36099 5 2 37834 4 2 38082 5 2 42404 5 2 42508 4 2 42810 4 2 45269 4 2 45493 5 2 46421 5 2 46977 3 2 47476 3 2 47478 2 2 47800 5 2 48482 5 2 48559 5 2 48687 1 2 49298 5 2 50096 5 2 50104 5 2 50133 5 2 51166 5 2 51460 4 2 51838 4 2 52036 3 2 53292 5 3 314 3 3 588 1 3 2077 2 3 2487 3 3 2900 3 3 3662 4 3 4536 1 3 5115 4 3 5272 5 3 5461 2 3 5885 4 3 6063 5 3 6630 5 3 6634 3 3 7001 4 3 9246 1 3 9731 2 3 9771 3 3 10140 3 3 10246 1 3 10249 5 3 10509 1 3 10610 1 3 10751 1 3 10944 1 3 11285 2 3 11408 3 3 11599 4 3 11854 1 3 11927 4 3 12471 2 3 12946 4 3 13274 5 3 13282 5 3 13794 1 3 14546 5 3 15604 1 3 16377 1 3 16569 1 3 16913 4 3 17434 4 3 17663 4 3 18031 3 3 18313 5 3 19724 4 3 19942 4 3 19984 2 3 20076 3 3 20782 4 3 20848 4 3 21228 3 3 21487 3 3 21733 1 3 22602 3 3 23576 1 3 23612 3 3 24326 3 3 24389 3 3 25182 3 3 25214 1 3 26398 2 3 26629 4 3 28767 3 3 29703 1 3 30681 4 3 32305 1 3 32592 5 3 32635 4 3 32918 5 3 33065 1 3 33716 3 3 36099 4 3 36240 2 3 37284 1 3 37834 4 3 37855 3 3 38080 5 3 38082 3 3 38475 3 3 42404 5 3 43602 4 3 43985 2 3 46977 2 3 47746 5 3 48482 3 3 48687 1 3 49288 5 3 49289 5 3 49295 5 3 49297 5 3 49298 1 3 50096 5 3 50101 5 3 50102 5 3 50104 5 3 50133 5 3 51166 3 3 51460 2 3 52036 1 3 53292 5 4 439 5 4 1169 5 4 2324 5 4 2487 5 4 3739 5 4 3922 4 4 4606 1 4 5115 5 4 5436 5 4 6342 5 4 6630 4 4 8167 5 4 8178 5 4 9246 4 4 10140 4 4 11285 3 4 11691 5 4 11692 5 4 11868 3 4 11945 5 4 12381 5 4 12874 5 4 12946 5 4 13034 4 4 13776 4 4 13794 4 4 14192 4 4 14207 3 4 14546 5 4 14603 5 4 15604 3 4 17228 5 4 17643 5 4 17663 5 4 17984 5 4 18031 5 4 18313 5 4 19526 5 4 19724 5 4 19729 4 4 20406 2 4 20467 4 4 21217 4 4 21228 3 4 21487 5 4 21676 3 4 21733 5 4 23576 5 4 24326 5 4 25164 5 4 25182 5 4 25840 3 4 26579 3 4 26629 4 4 26661 5 4 27499 5 4 27834 5 4 27934 4 4 28767 5 4 29123 3 4 29703 5 4 30283 4 4 30313 5 4 30681 5 4 30944 5 4 32305 5 4 32338 4 4 32592 5 4 32745 4 4 32748 5 4 32918 5 4 32923 5 4 33065 5 4 33697 5 4 33716 4 4 33872 5 4 34688 4 4 36099 5 4 37284 3 4 37834 4 4 39423 5 4 42404 5 4 43985 5 4 44243 3 4 44397 5 4 45269 5 4 46482 5 4 47411 4 4 47730 5 4 47800 5 4 48482 3 4 48559 5 4 48687 5 4 49298 5 4 50102 5 4 51460 5 4 51838 4 4 52036 3 4 53245 5 4 53292 5 5 314 4 5 1169 5 5 1952 5 5 2324 5 5 2487 3

In [6]:
#import books data
file_location = "/FileStore/tables/books.csv"
file_type = "csv"

# CSV options
infer_schema = "True"
first_row_is_header = "True"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
books = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sto_readep", delimiter) \
  .load(file_location)

display(books)

# Create Temporary Tables
books.createOrReplaceTempView("books")

id book_id best_book_id work_id books_count isbn isbn13 authors original_publication_year original_title title language_code average_rating ratings_count work_ratings_count work_text_reviews_count ratings_1 ratings_2 ratings_3 ratings_4 ratings_5 image_url small_image_url 1 2767052 2767052 2792775 272 439023483 9.78043902348E12 Suzanne Collins 2008.0 The Hunger Games The Hunger Games (The Hunger Games, #1) eng 4.34 4780653 4942365 155254 66715.0 127936 560092 1481305 2706317 https://images.gr-assets.com/books/1447303603m/2767052.jpg https://images.gr-assets.com/books/1447303603s/2767052.jpg 2 3 3 4640799 491 439554934 9.78043955493E12 J.K. Rowling, Mary GrandPré 1997.0 Harry Potter and the Philosopher's Stone Harry Potter and the Sorcerer's Stone (Harry Potter, #1) eng 4.44 4602479 4800065 75867 75504.0 101676 455024 1156318 3011543 https://images.gr-assets.com/books/1474154022m/3.jpg https://images.gr-assets.com/books/1474154022s/3.jpg 3 41865 41865 3212258 226 316015849 9.78031601584E12 Stephenie Meyer 2005.0 Twilight Twilight (Twilight, #1) en-US 3.57 3866839 3916824 95009 456191.0 436802 793319 875073 1355439 https://images.gr-assets.com/books/1361039443m/41865.jpg https://images.gr-assets.com/books/1361039443s/41865.jpg 4 2657 2657 3275794 487 61120081 9.78006112008E12 Harper Lee 1960.0 To Kill a Mockingbird To Kill a Mockingbird eng 4.25 3198671 3340896 72586 60427.0 117415 446835 1001952 1714267 https://images.gr-assets.com/books/1361975680m/2657.jpg https://images.gr-assets.com/books/1361975680s/2657.jpg 5 4671 4671 245494 1356 743273567 9.78074327356E12 F. Scott Fitzgerald 1925.0 The Great Gatsby The Great Gatsby eng 3.89 2683664 2773745 51992 86236.0 197621 606158 936012 947718 https://images.gr-assets.com/books/1490528560m/4671.jpg https://images.gr-assets.com/books/1490528560s/4671.jpg 6 11870085 11870085 16827462 226 525478817 9.78052547881E12 John Green 2012.0 The Fault in Our Stars The Fault in Our Stars eng 4.26 2346404 2478609 140739 47994.0 92723 327550 698471 1311871 https://images.gr-assets.com/books/1360206420m/11870085.jpg https://images.gr-assets.com/books/1360206420s/11870085.jpg 7 5907 5907 1540236 969 618260307 9.7806182603E12 J.R.R. Tolkien 1937.0 The Hobbit or There and Back Again The Hobbit en-US 4.25 2071616 2196809 37653 46023.0 76784 288649 665635 1119718 https://images.gr-assets.com/books/1372847500m/5907.jpg https://images.gr-assets.com/books/1372847500s/5907.jpg 8 5107 5107 3036731 360 316769177 9.78031676917E12 J.D. Salinger 1951.0 The Catcher in the Rye The Catcher in the Rye eng 3.79 2044241 2120637 44920 109383.0 185520 455042 661516 709176 https://images.gr-assets.com/books/1398034300m/5107.jpg https://images.gr-assets.com/books/1398034300s/5107.jpg 9 960 960 3338963 311 1416524797 9.78141652479E12 Dan Brown 2000.0 Angels & Demons Angels & Demons (Robert Langdon, #1) en-CA 3.85 2001311 2078754 25112 77841.0 145740 458429 716569 680175 https://images.gr-assets.com/books/1303390735m/960.jpg https://images.gr-assets.com/books/1303390735s/960.jpg 10 1885 1885 3060926 3455 679783261 9.78067978327E12 Jane Austen 1813.0 Pride and Prejudice Pride and Prejudice eng 4.24 2035490 2191465 49152 54700.0 86485 284852 609755 1155673 https://images.gr-assets.com/books/1320399351m/1885.jpg https://images.gr-assets.com/books/1320399351s/1885.jpg 11 77203 77203 3295919 283 1594480001 9.78159448E12 Khaled Hosseini 2003.0 The Kite Runner The Kite Runner eng 4.26 1813044 1878095 59730 34288.0 59980 226062 628174 929591 https://images.gr-assets.com/books/1484565687m/77203.jpg https://images.gr-assets.com/books/1484565687s/77203.jpg 12 13335037 13335037 13155899 210 62024035 9.78006202404E12 Veronica Roth 2011.0 Divergent Divergent (Divergent, #1) eng 4.24 1903563 2216814 101023 36315.0 82870 310297 673028 1114304 https://images.gr-assets.com/books/1328559506m/13335037.jpg https://images.gr-assets.com/books/1328559506s/13335037.jpg 13 5470 5470 153313 995 451524934 9.78045152494E12 George Orwell, Erich Fromm, C

In [7]:
#converting books data into pandas dataframe

books_df = books.toPandas()
books_df.head()

id 
 book_id 
 best_book_id 
 work_id 
 books_count 
 isbn 
 isbn13 
 authors 
 original_publication_year 
 original_title 
 title 
 language_code 
 average_rating 
 ratings_count 
 work_ratings_count 
 work_text_reviews_count 
 ratings_1 
 ratings_2 
 ratings_3 
 ratings_4 
 ratings_5 
 image_url 
 small_image_url 
 
 
 
 
 0 
 1 
 2767052 
 2767052 
 2792775 
 272 
 439023483 
 9.780439e+12 
 Suzanne Collins 
 2008.0 
 The Hunger Games 
 The Hunger Games (The Hunger Games, #1) 
 eng 
 4.34 
 4780653 
 4942365 
 155254 
 66715.0 
 127936 
 560092 
 1481305 
 2706317 
 https://images.gr-assets.com/books/1447303603m... 
 https://images.gr-assets.com/books/1447303603s... 
 
 
 1 
 2 
 3 
 3 
 4640799 
 491 
 439554934 
 9.780440e+12 
 J.K. Rowling, Mary GrandPré 
 1997.0 
 Harry Potter and the Philosopher's Stone 
 Harry Potter and the Sorcerer's Stone (Harry P... 
 eng 
 4.44 
 4602479 
 4800065 
 75867 
 75504.0 
 101676 
 455024 
 1156318 
 3011543 
 https://images.gr-assets.com/books/1474154022m... 
 https://images.gr-assets.com/books/1474154022s... 
 
 
 2 
 3 
 41865 
 41865 
 3212258 
 226 
 316015849 
 9.780316e+12 
 Stephenie Meyer 
 2005.0 
 Twilight 
 Twilight (Twilight, #1) 
 en-US 
 3.57 
 3866839 
 3916824 
 95009 
 456191.0 
 436802 
 793319 
 875073 
 1355439 
 https://images.gr-assets.com/books/1361039443m... 
 https://images.gr-assets.com/books/1361039443s... 
 
 
 3 
 4 
 2657 
 2657 
 3275794 
 487 
 61120081 
 9.780061e+12 
 Harper Lee 
 1960.0 
 To Kill a Mockingbird 
 To Kill a Mockingbird 
 eng 
 4.25 
 3198671 
 3340896 
 72586 
 60427.0 
 117415 
 446835 
 1001952 
 1714267 
 https://images.gr-assets.com/books/1361975680m... 
 https://images.gr-assets.com/books/1361975680s... 
 
 
 4 
 5 
 4671 
 4671 
 245494 
 1356 
 743273567 
 9.780743e+12 
 F. Scott Fitzgerald 
 1925.0 
 The Great Gatsby 
 The Great Gatsby 
 eng 
 3.89 
 2683664 
 2773745 
 51992 
 86236.0 
 197621 
 606158 
 936012 
 947718 
 https://images.gr-assets.com/books/1490528560m... 
 https://images.gr-assets.com/books/1490528560s...

There is a book from 1750BC

In [9]:
%sql

select * from books where original_publication_year = -1750

id book_id best_book_id work_id books_count isbn isbn13 authors original_publication_year original_title title language_code average_rating ratings_count work_ratings_count work_text_reviews_count ratings_1 ratings_2 ratings_3 ratings_4 ratings_5 image_url small_image_url 2076 19351 19351 3802528 266 141026286 9.78014102628E12 Anonymous, N.K. Sandars -1750.0 Shūtur eli sharrī The Epic of Gilgamesh eng 3.63 44345 55856 2247 1551.0 5850 17627 17485 13343 https://s.gr-assets.com/assets/nophoto/book/111x148-bcc042a9c91a29c1d680899eff700a03.png https://s.gr-assets.com/assets/nophoto/book/50x75-a91bf249278a81aabab721ef782c4a74.png

**Distribution of numberr of books published across years**

In [11]:
%sql

select original_publication_year, count(*) as count from books where original_publication_year > 1950 group by original_publication_year 

original_publication_year count 1988.0 89 1951.0 13 1976.0 39 1979.0 48 1953.0 26 1987.0 83 1959.0 24 1978.0 48 1968.0 35 2010.0 473 1967.0 24 1964.0 19 1993.0 115 2001.0 226 1954.0 21 1965.0 23 1984.0 67 1973.0 35 1980.0 54 1966.0 21 1997.0 168 1992.0 112 1990.0 99 1995.0 126 2009.0 432 2007.0 363 1960.0 22 1996.0 163 1963.0 31 1969.0 32 1986.0 78 1998.0 172 1956.0 26 1985.0 77 2017.0 11 1957.0 19 1982.0 75 2000.0 209 1994.0 121 1970.0 33 2008.0 383 1952.0 19 1999.0 186 1981.0 50 2011.0 556 1975.0 40 1989.0 87 2002.0 225 1958.0 24 1962.0 32 2006.0 362 1961.0 33 2004.0 307 2013.0 518 1955.0 24 2012.0 568 1972.0 30 1991.0 98 1983.0 44 1977.0 56 1974.0 40 2014.0 437 2015.0 306 2005.0 326 2003.0 288 2016.0 198 1971.0 31

In [12]:
import pandas as pd
import numpy as np

In [13]:
books_df.shape

Out[7]: (10000, 23)

In [14]:
3sanity check
books_df = books_df[books_df.ratings_count.str.isdigit() == True]

In [15]:
books_df.ratings_count = books_df.ratings_count.astype('int')

In [16]:
#Top books with most number of ratings on goodbooks

books_df.sort_values(by = 'ratings_count', ascending = False)[['original_title','ratings_count', 'average_rating' ]][0:10]

original_title 
 ratings_count 
 average_rating 
 
 
 
 
 0 
 The Hunger Games 
 4780653 
 4.34 
 
 
 1 
 Harry Potter and the Philosopher's Stone 
 4602479 
 4.44 
 
 
 2 
 Twilight 
 3866839 
 3.57 
 
 
 3 
 To Kill a Mockingbird 
 3198671 
 4.25 
 
 
 4 
 The Great Gatsby 
 2683664 
 3.89 
 
 
 5 
 The Fault in Our Stars 
 2346404 
 4.26 
 
 
 6 
 The Hobbit or There and Back Again 
 2071616 
 4.25 
 
 
 7 
 The Catcher in the Rye 
 2044241 
 3.79 
 
 
 9 
 Pride and Prejudice 
 2035490 
 4.24 
 
 
 8 
 Angels & Demons 
 2001311 
 3.85

In [17]:
most_ratings = books_df.sort_values(by = 'ratings_count', ascending = False)[['original_title','ratings_count', 'average_rating', 'image_url' ]][0:10]

**Printing top books with most number of ratings**

In [19]:
import pandas as pd
from IPython.display import Image, HTML
most_ratings['img_html'] = most_ratings['image_url']\
    .str.replace(
        '(.*)', 
        '<img src="\\1" style="max-height:124px;"></img>'
    )
with pd.option_context('display.max_colwidth', 10000):
  
  display(HTML(most_ratings[['original_title', 'img_html', 'ratings_count', 'average_rating' ]].to_html(escape=False)))

original_title 
 img_html 
 ratings_count 
 average_rating 
 
 
 
 
 0 
 The Hunger Games 
 
 4780653 
 4.34 
 
 
 1 
 Harry Potter and the Philosopher's Stone 
 
 4602479 
 4.44 
 
 
 2 
 Twilight 
 
 3866839 
 3.57 
 
 
 3 
 To Kill a Mockingbird 
 
 3198671 
 4.25 
 
 
 4 
 The Great Gatsby 
 
 2683664 
 3.89 
 
 
 5 
 The Fault in Our Stars 
 
 2346404 
 4.26 
 
 
 6 
 The Hobbit or There and Back Again 
 
 2071616 
 4.25 
 
 
 7 
 The Catcher in the Rye 
 
 2044241 
 3.79 
 
 
 9 
 Pride and Prejudice 
 
 2035490 
 4.24 
 
 
 8 
 Angels & Demons 
 
 2001311 
 3.85

In [20]:
books_df.head()

id 
 book_id 
 best_book_id 
 work_id 
 books_count 
 isbn 
 isbn13 
 authors 
 original_publication_year 
 original_title 
 title 
 language_code 
 average_rating 
 ratings_count 
 work_ratings_count 
 work_text_reviews_count 
 ratings_1 
 ratings_2 
 ratings_3 
 ratings_4 
 ratings_5 
 image_url 
 small_image_url 
 
 
 
 
 0 
 1 
 2767052 
 2767052 
 2792775 
 272 
 439023483 
 9.780439e+12 
 Suzanne Collins 
 2008.0 
 The Hunger Games 
 The Hunger Games (The Hunger Games, #1) 
 eng 
 4.34 
 4780653 
 4942365 
 155254 
 66715.0 
 127936 
 560092 
 1481305 
 2706317 
 https://images.gr-assets.com/books/1447303603m... 
 https://images.gr-assets.com/books/1447303603s... 
 
 
 1 
 2 
 3 
 3 
 4640799 
 491 
 439554934 
 9.780440e+12 
 J.K. Rowling, Mary GrandPré 
 1997.0 
 Harry Potter and the Philosopher's Stone 
 Harry Potter and the Sorcerer's Stone (Harry P... 
 eng 
 4.44 
 4602479 
 4800065 
 75867 
 75504.0 
 101676 
 455024 
 1156318 
 3011543 
 https://images.gr-assets.com/books/1474154022m... 
 https://images.gr-assets.com/books/1474154022s... 
 
 
 2 
 3 
 41865 
 41865 
 3212258 
 226 
 316015849 
 9.780316e+12 
 Stephenie Meyer 
 2005.0 
 Twilight 
 Twilight (Twilight, #1) 
 en-US 
 3.57 
 3866839 
 3916824 
 95009 
 456191.0 
 436802 
 793319 
 875073 
 1355439 
 https://images.gr-assets.com/books/1361039443m... 
 https://images.gr-assets.com/books/1361039443s... 
 
 
 3 
 4 
 2657 
 2657 
 3275794 
 487 
 61120081 
 9.780061e+12 
 Harper Lee 
 1960.0 
 To Kill a Mockingbird 
 To Kill a Mockingbird 
 eng 
 4.25 
 3198671 
 3340896 
 72586 
 60427.0 
 117415 
 446835 
 1001952 
 1714267 
 https://images.gr-assets.com/books/1361975680m... 
 https://images.gr-assets.com/books/1361975680s... 
 
 
 4 
 5 
 4671 
 4671 
 245494 
 1356 
 743273567 
 9.780743e+12 
 F. Scott Fitzgerald 
 1925.0 
 The Great Gatsby 
 The Great Gatsby 
 eng 
 3.89 
 2683664 
 2773745 
 51992 
 86236.0 
 197621 
 606158 
 936012 
 947718 
 https://images.gr-assets.com/books/1490528560m... 
 https://images.gr-assets.com/books/1490528560s...

In [21]:
books_df.average_rating = books_df.average_rating.astype('float')

In [22]:
#Top books with top average ratings on goodbooks

books_df.sort_values(by = 'average_rating', ascending = False)[['original_title','ratings_count', 'average_rating' ]][0:10]

original_title 
 ratings_count 
 average_rating 
 
 
 
 
 3627 
 The Complete Calvin and Hobbes 
 28900 
 4.82 
 
 
 861 
 Words of Radiance 
 73572 
 4.77 
 
 
 3274 
 None 
 33220 
 4.77 
 
 
 8853 
 Mark of the Lion Trilogy 
 9081 
 4.76 
 
 
 7946 
 None 
 8953 
 4.76 
 
 
 4482 
 It's a Magical World: A Calvin and Hobbes Coll... 
 22351 
 4.75 
 
 
 6360 
 There's Treasure Everywhere: A Calvin and Hobb... 
 16766 
 4.74 
 
 
 421 
 Complete Harry Potter Boxed Set 
 190050 
 4.74 
 
 
 6589 
 The Authoritative Calvin and Hobbes 
 16087 
 4.73 
 
 
 6919 
 The Indispensable Calvin and Hobbes: A Calvin ... 
 14597 
 4.73

In [23]:
high_rating_books = books_df.sort_values(by = 'average_rating', ascending = False)[['original_title','ratings_count','image_url', 'average_rating' ]][0:10]

**Printing top books with highest average ratings**

In [25]:
high_rating_books['img_html'] = high_rating_books['image_url']\
    .str.replace(
        '(.*)', 
        '<img src="\\1" style="max-height:124px;"></img>'
    )
with pd.option_context('display.max_colwidth', 10000):
  
  display(HTML(high_rating_books[['original_title', 'img_html','ratings_count', 'average_rating' ]].to_html(escape=False)))

original_title 
 img_html 
 ratings_count 
 average_rating 
 
 
 
 
 3627 
 The Complete Calvin and Hobbes 
 
 28900 
 4.82 
 
 
 861 
 Words of Radiance 
 
 73572 
 4.77 
 
 
 3274 
 None 
 
 33220 
 4.77 
 
 
 8853 
 Mark of the Lion Trilogy 
 
 9081 
 4.76 
 
 
 7946 
 None 
 
 8953 
 4.76 
 
 
 4482 
 It's a Magical World: A Calvin and Hobbes Collection 
 
 22351 
 4.75 
 
 
 6360 
 There's Treasure Everywhere: A Calvin and Hobbes Collection 
 
 16766 
 4.74 
 
 
 421 
 Complete Harry Potter Boxed Set 
 
 190050 
 4.74 
 
 
 6589 
 The Authoritative Calvin and Hobbes 
 
 16087 
 4.73 
 
 
 6919 
 The Indispensable Calvin and Hobbes: A Calvin and Hobbes Treasury 
 
 14597 
 4.73

In [26]:
authors_with_most_books = pd.DataFrame(books_df.authors.value_counts()[0:10]).reset_index()
authors_with_most_books.columns = ['author', 'number_of_books']

In [27]:
 authors_with_most_books

author 
 number_of_books 
 
 
 
 
 0 
 Stephen King 
 60 
 
 
 1 
 Nora Roberts 
 59 
 
 
 2 
 Dean Koontz 
 47 
 
 
 3 
 Terry Pratchett 
 42 
 
 
 4 
 Agatha Christie 
 39 
 
 
 5 
 Meg Cabot 
 37 
 
 
 6 
 James Patterson 
 36 
 
 
 7 
 David Baldacci 
 34 
 
 
 8 
 John Grisham 
 33 
 
 
 9 
 J.D. Robb 
 33

In [28]:
import matplotlib.pyplot as plt
plt.figure(figsize=(12,6))
plt.title('Distribution of Average Ratings')
books_df['average_rating'].hist()
display()

In [29]:
ratings.show(5)

+-------+-------+------+
book_id|user_id|rating|
+-------+-------+------+
 1| 314| 5|
 1| 439| 3|
 1| 588| 5|
 1| 1169| 4|
 1| 1185| 4|
+-------+-------+------+
only showing top 5 rows

In [30]:
#1 million ratings

ratings.describe().show()

+-------+-----------------+------------------+------------------+
summary| book_id| user_id| rating|
+-------+-----------------+------------------+------------------+
 count| 981756| 981756| 981756|
 mean|4943.275635697668|25616.759933221696|3.8565335989797873|
 stddev|2873.207414896143|15228.338825882149|0.9839408559619973|
 min| 1| 1| 1|
 max| 10000| 53424| 5|
+-------+-----------------+------------------+------------------+

In [31]:
#Each book has 100 ratings in the ratingss dataframe

ratings.groupby('book_id').count().show()

+-------+-----+
book_id|count|
+-------+-----+
 148| 100|
 463| 100|
 471| 100|
 496| 100|
 833| 100|
 1088| 100|
 1238| 100|
 1342| 100|
 1580| 100|
 1591| 100|
 1645| 100|
 1829| 100|
 1959| 100|
 2122| 100|
 2142| 100|
 2366| 100|
 2659| 100|
 2866| 100|
 3175| 100|
 3749| 100|
+-------+-----+
only showing top 20 rows

Here's the few books with less than **100 ratings**

In [33]:
%sql

SELECT b.original_title, r.book_id,count(*)  FROM ratings r inner join books b on b.book_id = r.book_id group by r.book_id, b.original_title having count(*) <100 

original_title book_id count(1) I Can Read with My Eyes Shut! 7785 98 Fried Green Tomatoes at the Whistle Stop Cafe 9375 77 Bluebeard 9601 87 Papillon 6882 95 A Case of Need 7663 91 The Complete Short Stories of Ernest Hemingway 4625 96 Trials of Death (Cirque du Freak, #5) 8967 89 The Vampire Prince (Cirque Du Freak, #6) 8968 96 Comfort Me with Apples: More Adventures at the Table 8725 99 El club Dumas 7194 95 Mariel of Redwall 7993 99 Still Life with Woodpecker 9566 99 number9dream 6820 98 The Man Who Listens to Horses 8515 97 Emma 6969 91 The Good Soldier 7628 85 Maximum City: Bombay Lost and Found 4364 89 How to Make Love Like a Porn Star: A Cautionary Tale 7252 99 Anne of Green Gables 8127 99 Shadow of the Hegemon 9534 57 Superman: Birthright 1-12 8172 95 Confessions of a Shopaholic 9416 71 Timequake 9594 98 The Mermaid Chair 6976 92 Eaters of the Dead 7673 99 Breakfast of Champions 4980 98 The Smoke Jumper 9282 97 Hard Times: For These Times 5344 99 null 9522 90 Lolita 7604 99 The First Four Years 8296 91 Genome: The Autobiography of a Species in 23 Chapters 4591 91 'Salem's Lot The Illustrated Edition 5413 74 Reservation Blues 6159 99 One Fish, Two Fish, Red Fish, Blue Fish 7770 98 The God of Small Things 9777 98 Path of the Assassin 7520 99 Xenocide 8648 94 Saving Fish from Drowning: A Novel 9503 97 Prep 9844 93 The Tragedy of Macbeth 8852 99 Less Than Zero 9915 90 The Informers 9914 94 Last Orders 5068 98 Even Cowgirls Get the Blues 7572 99 I Feel Bad About My Neck: And Other Thoughts on Being a Woman 8765 97 Sons of Destiny (Cirque Du Freak, Book 12) 8948 61 Another Roadside Attraction 9570 83 As The Crow Flies 7201 97 Freak the Mighty 8953 97 Witches Abroad 2442 98 A Woman of Substance 8155 99 The Atonement Child 9943 89 Rachel's Holiday 9301 78 Ghostwritten 6819 60 The Idiot Girls' Action-Adventure Club: True Tales from a Magnificent and Clumsy Life 7531 99 Marlfox 7999 98 The Bourne Identity 7869 91 Visions of Sugar Plums 6420 98 Half Asleep in Frog Pajamas 9567 94 Atonement 6867 93 The Red Pony 8732 86 Shopaholic Ties the Knot 9419 89 Suzanne's Diary for Nicholas 7494 89 The Electric Kool-Aid Acid Test 7442 99 Don't Make Me Think! A common sense approach to web usability 3368 97 Prophet 8492 99 The Restaurant at the End of the Universe 8695 99 American Gods 4407 99 The Lake of Souls (Cirque Du Freak, #10) 8951 99 Morality for Beautiful Girls 7035 99 Four Blondes 6613 93 Allies of the Night (Cirque Du Freak, Book 8) 8966 98 La casa de los espíritus 9328 84 The Audacity of Hope: Thoughts on Reclaiming the American Dream 9742 86 BLEACH―ブリーチ―　1 2880 99 Jailbird 4987 99 The Naked Lunch 7437 91 Fear of Flying 9654 97 Il barone rampante 9804 88 Animal Farm & 1984 5472 99 Last Chance Saloon 9298 97 Pardonable Lies 7432 96 Desert Flower: The Extraordinary Journey of a Desert Nomad 8745 92 From Potter's Field 6537 97 Forever in Blue: The Fourth Summer of the Sisterhood 5453 96 Horton Hears a Who! 7779 97 Girls in Pants: The Third Summer of the Sisterhood 5452 99 Преступление и наказание 7144 76 Jurassic Park 7677 69 Homage to Catalonia 9646 75 In the Lake of the Woods 3447 99 Reading Lolita in Tehran: A Memoir in Books 7603 99 The Fortress of Solitude 9799 97 Predator 6528 98 King Rat 9827 96 The Broom of the System 6750 99 Player Piano 9597 95 Suffer the Children 6572 76 Awakening the Buddha Within : Tibetan Wisdom for the Western World 7579 99 World Without End 5064 95 Cirque Du Freak (Cirque du Freak, #1) 8962 86 The Ground Beneath Her Feet 9864 94 Amsterdam 6862 95 Persepolis 2 9517 87 Mother Night 9592 99 Les Trois Mousquetaires 7190 99 Watermelon 9300 95 Die Verwandlung und andere Erzählungen 7723 99 The Joy Luck Club 7763 87 Hunters of the Dusk (Cirque Du Freak, #7) 8963 99 Shopaholic and Baby 9420 90 How to Stop Worrying and Start Living 4866 95 Keep the Aspidistra Flying 9648 91 Motor Mouth 6426 97 All-of-a-Kind Family 7926 99 El amor en los tiempos del cólera 9712 94 Slapstick 9595 93 The Andromeda Strain 7670 99

In [34]:
ratings = ratings.select(ratings.user_id,
                         ratings.book_id,
                         ratings.rating.cast("double"))

In [35]:
ratings.show(5)

+-------+-------+------+
user_id|book_id|rating|
+-------+-------+------+
 314| 1| 5.0|
 439| 1| 3.0|
 588| 1| 5.0|
 1169| 1| 4.0|
 1185| 1| 4.0|
+-------+-------+------+
only showing top 5 rows

In [36]:
# Count the total number of ratings in the dataset
numerator = ratings.select("rating").count()

# Count the number of distinct Id's
num_users = ratings.select("user_id").distinct().count()
num_items = ratings.select("book_id").distinct().count()

# Set the denominator equal to the number of users multiplied by the number of items
denominator = num_users * num_items

# Divide the numerator by the denominator
sparsity = (1.0 - (numerator * 1.0)/ denominator) * 100
print("The ratings dataframe is ", "%.2f" % sparsity + "% empty.")

The ratings dataframe is 99.82% empty.

In [37]:
# Min num ratings 
print("Item with the fewest ratings: ")
ratings.groupBy("book_id").count().sort('count').show(10)

Item with the fewest ratings: 
+-------+-----+
book_id|count|
+-------+-----+
 7803| 8|
 9345| 11|
 9486| 24|
 1935| 34|
 9315| 36|
 9553| 41|
 8181| 44|
 9260| 46|
 9479| 47|
 9705| 48|
+-------+-----+
only showing top 10 rows

In [38]:
# Group data by user_id, count ratings
(ratings.groupBy("user_id")
    .count()
    .filter("`count` >= 5")
    .orderBy('count', ascending=False)
    .show(n = 10))

+-------+-----+
user_id|count|
+-------+-----+
 12874| 200|
 30944| 200|
 28158| 199|
 12381| 199|
 52036| 199|
 6630| 197|
 37834| 197|
 45554| 197|
 24143| 196|
 9668| 196|
+-------+-----+
only showing top 10 rows

In [39]:
(ratings.groupBy("book_id")
    .count()
    .filter("`count` > 1")
    .orderBy('count', ascending=False)
    .show(n = 10))

+-------+-----+
book_id|count|
+-------+-----+
 31| 100|
 516| 100|
 1139| 100|
 1143| 100|
 1270| 100|
 1303| 100|
 1322| 100|
 1339| 100|
 1352| 100|
 1618| 100|
+-------+-----+
only showing top 10 rows

**Now, let's split the data into training and test set to use collaborative filtering usingn Alternate Least Square method**

In [41]:
(training, test) = ratings.randomSplit([0.8, 0.2])

In [42]:
test.show(5)

+-------+-------+------+
user_id|book_id|rating|
+-------+-------+------+
 7| 1646| 3.0|
 7| 2045| 3.0|
 7| 2129| 5.0|
 7| 2189| 3.0|
 7| 2325| 4.0|
+-------+-------+------+
only showing top 5 rows

**Let's import ALS and regression evaluator to find RMSE.**

In [44]:
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.recommendation import ALS
from pyspark.sql import Row

In [45]:
als = ALS( userCol="user_id", itemCol="book_id", ratingCol="rating",
          coldStartStrategy="drop", nonnegative = True, implicitPrefs = False)

**'implicitPrefs ' argument is used when we are not using explicit data like the rating data here. Sometimes companies don't have explicit data like ratings and still want to build a recommendation engine using other proxies like views, clicks, wishlists, etc. **

**In that case, implicit preference is used but is out of the scope of our good books project. 'coldStartStrategy ' is used when we don't have any data for a user which might lead to null prediction if the user on the test set has no rating in the training set. We have dropped the cold start strategy because we want to avoid such situations for our problem in hand.**


**Now let's build our hyperparameter list and then fit the algorithm on the training data.**

In [47]:
type(als)

Out[42]: pyspark.ml.recommendation.ALS

In [48]:
from pyspark.ml.tuning import ParamGridBuilder, CrossValidator 

param_grid = ParamGridBuilder() \
            .addGrid(als.rank, [10, 50, 75, 100]) \
            .addGrid(als.maxIter, [5, 50, 75, 100]) \
            .addGrid(als.regParam, [.01, .05, .1, .15]) \
            .build()

**Num models to be tested using param_grid: 64
There is a total of 64 models that will be tested and tuned before we receive the final model. The power of parallelization and spark will make it pretty fast though.**

In [50]:
# Define evaluator as RMSE
evaluator = RegressionEvaluator(metricName = "rmse", 
                                labelCol = "rating", 
                                predictionCol = "prediction")
# Print length of evaluator
print ("Num models to be tested using param_grid: ", len(param_grid))

Num models to be tested using param_grid: 64

In [51]:
# Build cross validation using CrossValidator
cv = CrossValidator(estimator = als, 
                    estimatorParamMaps = param_grid, 
                    evaluator = evaluator, 
                    numFolds = 5)

In [52]:
print(cv)

CrossValidator_089d3e0e8add

In [53]:
model = als.fit(training)

In [54]:
predictions = model.transform(test)

In [55]:
predictions.show(n = 10)

+-------+-------+------+----------+
user_id|book_id|rating|prediction|
+-------+-------+------+----------+
 3922| 148| 3.0| 3.7151182|
 32055| 148| 3.0| 3.1291625|
 27834| 148| 3.0| 3.832256|
 13407| 148| 4.0| 3.9250367|
 29703| 148| 4.0| 3.7403727|
 40167| 148| 5.0| 3.2674868|
 46139| 148| 5.0| 3.8008392|
 17228| 148| 5.0| 3.9937172|
 14372| 148| 3.0| 3.188208|
 30681| 148| 2.0| 3.1382155|
+-------+-------+------+----------+
only showing top 10 rows

**Well, the prediction on the test set shows that it is very close to the original rating. Like for example the rating for user_id 14372 was originally 3 and our algorithm predicted it to be 3.13 which is pretty close.**

In [57]:
predictions.createOrReplaceTempView("predictions")

In [58]:
%sql
select * from predictions

user_id book_id rating prediction 3922 148 3.0 3.7151182 32055 148 3.0 3.1291625 27834 148 3.0 3.832256 13407 148 4.0 3.9250367 29703 148 4.0 3.7403727 40167 148 5.0 3.2674868 46139 148 5.0 3.8008392 17228 148 5.0 3.9937172 14372 148 3.0 3.188208 30681 148 2.0 3.1382155 10944 148 4.0 3.747405 11945 148 4.0 3.9951036 45811 148 3.0 3.7638686 6689 463 4.0 3.353484 5152 463 4.0 4.221477 41145 463 4.0 4.0171585 1294 463 4.0 3.928851 12806 463 3.0 4.162929 52012 463 5.0 3.359523 8043 463 4.0 3.9729245 16081 463 4.0 4.649987 50999 463 5.0 4.61083 48493 463 5.0 4.3947697 1947 463 4.0 4.133591 11191 463 5.0 4.740238 22154 463 3.0 3.5734375 28540 463 5.0 4.7005625 2271 463 3.0 3.535572 9195 463 3.0 3.3095796 11502 463 2.0 3.472526 26146 463 5.0 4.6619616 14599 463 4.0 4.827334 52334 463 5.0 4.8373957 20782 463 3.0 4.265115 11868 471 5.0 4.0514975 24326 471 4.0 3.9500256 32055 471 3.0 3.4907005 1456 471 5.0 4.197986 14890 471 4.0 3.1573334 18361 471 5.0 4.0132713 27934 471 2.0 3.595066 9668 471 5.0 3.6639752 30184 471 5.0 4.9322796 45554 471 4.0 3.8707287 1677 471 4.0 3.7522612 21543 471 3.0 3.2413416 3662 471 3.0 4.049785 30879 471 3.0 3.3625264 15054 471 4.0 3.3142824 17228 471 5.0 4.430365 23004 471 5.0 4.5990214 3641 471 5.0 3.9084647 22602 471 3.0 3.3971052 21658 471 4.0 3.760411 10335 471 4.0 3.7184696 9165 496 3.0 3.7442455 12466 496 2.0 4.0919433 6630 496 4.0 3.362639 11927 496 5.0 3.489345 16576 496 5.0 3.996826 13407 496 5.0 4.1826878 15889 496 4.0 3.5064974 29031 496 3.0 3.2273057 17359 496 4.0 3.959391 9125 496 1.0 2.1962235 16569 496 2.0 4.0398126 30304 496 4.0 3.5191643 18550 496 4.0 4.0106063 10402 496 4.0 3.3780074 9804 496 5.0 3.689182 22602 496 3.0 3.3498814 41747 496 3.0 3.4196923 32576 496 5.0 3.6654968 49514 496 4.0 3.5582395 14372 496 4.0 3.4481738 8370 496 3.0 3.6750364 11837 496 4.0 3.1264987 9244 496 5.0 3.6416903 15494 496 4.0 2.945429 9843 496 5.0 4.5137606 1296 833 3.0 3.6615472 25616 833 2.0 2.6013682 31615 833 3.0 3.850278 37041 833 3.0 2.9470046 26750 833 5.0 4.349528 4212 833 4.0 3.432973 17643 833 3.0 3.9918232 30184 833 5.0 4.5021844 6567 833 4.0 3.0734699 49152 833 4.0 3.5545359 51838 833 4.0 3.5363493 48196 833 3.0 2.9169917 29819 833 3.0 3.521237 9285 833 4.0 3.8349485 46977 833 3.0 2.9991443 23612 833 2.0 2.7155385 24834 833 5.0 4.0947638 9843 833 5.0 4.5258946 32305 1088 2.0 3.0898216 11868 1088 4.0 3.6575704 25606 1088 4.0 2.653628 22164 1088 3.0 3.3920646 29732 1088 5.0 3.8353536 15186 1088 3.0 2.9232552 20229 1088 5.0 4.20609 20241 1088 4.0 4.0397224 36695 1088 3.0 3.7512681 53012 1088 5.0 4.5396614 5645 1088 5.0 4.428069 14839 1088 3.0 3.9327407 49850 1088 5.0 4.4653306 40354 1088 3.0 4.606335 49988 1088 3.0 3.3525014 32574 1088 5.0 4.9283724 5695 1238 3.0 2.8628404 1533 1238 4.0 4.015832 41300 1238 4.0 3.786705 10309 1238 3.0 3.301256 1565 1238 4.0 3.8240943 33065 1238 4.0 3.615452 18620 1238 4.0 3.2935855 7523 1238 3.0 4.1331 28694 1238 5.0 4.051745 48726 1238 3.0 4.019161 6829 1238 5.0 3.1146653 2918 1238 4.0 3.5618095 39032 1238 5.0 3.724786 5629 1238 4.0 3.7003121 13962 1238 5.0 3.7371306 14543 1238 4.0 3.623846 27279 1238 2.0 2.4887006 45269 1238 3.0 3.1725745 7955 1238 5.0 4.247321 4795 1238 5.0 3.9672122 6460 1342 5.0 4.1374683 24225 1342 4.0 4.0180855 47249 1342 3.0 3.6206627 14681 1342 5.0 4.875955 41354 1342 5.0 4.7369866 23768 1342 4.0 4.1130595 20780 1342 4.0 4.113356 36088 1342 5.0 4.458716 38067 1342 4.0 4.460821 19378 1342 5.0 4.046038 41249 1342 5.0 4.36773 40700 1342 5.0 4.27321 40841 1342 5.0 4.3429728 43602 1342 5.0 4.5349665 26718 1342 4.0 3.936307 25988 1342 5.0 4.51884 40201 1342 5.0 5.1048746 40468 1342 5.0 4.735393 15123 1342 4.0 4.57669 45116 1342 3.0 4.242945 1088 1580 5.0 3.4965446 47018 1580 4.0 3.7935514 8519 1580 5.0 3.589721 11502 1580 4.0 3.489625 43371 1580 4.0 3.1111042 5267 1580 4.0 3.537835 37284 1580 3.0 3.6288385 33633 1580 3.0 3.3092957 46472 1580 4.0 3.3758793 18556 1580 3.0 3.9775994 40354 1580 5.0 3.868647 16417 1580 3.0 3.5014403 35521 1580 3.0 3.60692

In [59]:
%sql
select predictions.user_id, predictions.book_id, predictions.rating, predictions.prediction, books.title from 
predictions inner join books 
ON predictions.book_id = books.id

user_id book_id rating prediction title 3922 148 3.0 3.7151182 Girl with a Pearl Earring 32055 148 3.0 3.1291625 Girl with a Pearl Earring 27834 148 3.0 3.832256 Girl with a Pearl Earring 13407 148 4.0 3.9250367 Girl with a Pearl Earring 29703 148 4.0 3.7403727 Girl with a Pearl Earring 40167 148 5.0 3.2674868 Girl with a Pearl Earring 46139 148 5.0 3.8008392 Girl with a Pearl Earring 17228 148 5.0 3.9937172 Girl with a Pearl Earring 14372 148 3.0 3.188208 Girl with a Pearl Earring 30681 148 2.0 3.1382155 Girl with a Pearl Earring 10944 148 4.0 3.747405 Girl with a Pearl Earring 11945 148 4.0 3.9951036 Girl with a Pearl Earring 45811 148 3.0 3.7638686 Girl with a Pearl Earring 6689 463 4.0 3.353484 Dragonfly in Amber (Outlander, #2) 5152 463 4.0 4.221477 Dragonfly in Amber (Outlander, #2) 41145 463 4.0 4.0171585 Dragonfly in Amber (Outlander, #2) 1294 463 4.0 3.928851 Dragonfly in Amber (Outlander, #2) 12806 463 3.0 4.162929 Dragonfly in Amber (Outlander, #2) 52012 463 5.0 3.359523 Dragonfly in Amber (Outlander, #2) 8043 463 4.0 3.9729245 Dragonfly in Amber (Outlander, #2) 16081 463 4.0 4.649987 Dragonfly in Amber (Outlander, #2) 50999 463 5.0 4.61083 Dragonfly in Amber (Outlander, #2) 48493 463 5.0 4.3947697 Dragonfly in Amber (Outlander, #2) 1947 463 4.0 4.133591 Dragonfly in Amber (Outlander, #2) 11191 463 5.0 4.740238 Dragonfly in Amber (Outlander, #2) 22154 463 3.0 3.5734375 Dragonfly in Amber (Outlander, #2) 28540 463 5.0 4.7005625 Dragonfly in Amber (Outlander, #2) 2271 463 3.0 3.535572 Dragonfly in Amber (Outlander, #2) 9195 463 3.0 3.3095796 Dragonfly in Amber (Outlander, #2) 11502 463 2.0 3.472526 Dragonfly in Amber (Outlander, #2) 26146 463 5.0 4.6619616 Dragonfly in Amber (Outlander, #2) 14599 463 4.0 4.827334 Dragonfly in Amber (Outlander, #2) 52334 463 5.0 4.8373957 Dragonfly in Amber (Outlander, #2) 20782 463 3.0 4.265115 Dragonfly in Amber (Outlander, #2) 11868 471 5.0 4.0514975 Mansfield Park 24326 471 4.0 3.9500256 Mansfield Park 32055 471 3.0 3.4907005 Mansfield Park 1456 471 5.0 4.197986 Mansfield Park 14890 471 4.0 3.1573334 Mansfield Park 18361 471 5.0 4.0132713 Mansfield Park 27934 471 2.0 3.595066 Mansfield Park 9668 471 5.0 3.6639752 Mansfield Park 30184 471 5.0 4.9322796 Mansfield Park 45554 471 4.0 3.8707287 Mansfield Park 1677 471 4.0 3.7522612 Mansfield Park 21543 471 3.0 3.2413416 Mansfield Park 3662 471 3.0 4.049785 Mansfield Park 30879 471 3.0 3.3625264 Mansfield Park 15054 471 4.0 3.3142824 Mansfield Park 17228 471 5.0 4.430365 Mansfield Park 23004 471 5.0 4.5990214 Mansfield Park 3641 471 5.0 3.9084647 Mansfield Park 22602 471 3.0 3.3971052 Mansfield Park 21658 471 4.0 3.760411 Mansfield Park 10335 471 4.0 3.7184696 Mansfield Park 9165 496 3.0 3.7442455 The Clan of the Cave Bear (Earth's Children, #1) 12466 496 2.0 4.0919433 The Clan of the Cave Bear (Earth's Children, #1) 6630 496 4.0 3.362639 The Clan of the Cave Bear (Earth's Children, #1) 11927 496 5.0 3.489345 The Clan of the Cave Bear (Earth's Children, #1) 16576 496 5.0 3.996826 The Clan of the Cave Bear (Earth's Children, #1) 13407 496 5.0 4.1826878 The Clan of the Cave Bear (Earth's Children, #1) 15889 496 4.0 3.5064974 The Clan of the Cave Bear (Earth's Children, #1) 29031 496 3.0 3.2273057 The Clan of the Cave Bear (Earth's Children, #1) 17359 496 4.0 3.959391 The Clan of the Cave Bear (Earth's Children, #1) 9125 496 1.0 2.1962235 The Clan of the Cave Bear (Earth's Children, #1) 16569 496 2.0 4.0398126 The Clan of the Cave Bear (Earth's Children, #1) 30304 496 4.0 3.5191643 The Clan of the Cave Bear (Earth's Children, #1) 18550 496 4.0 4.0106063 The Clan of the Cave Bear (Earth's Children, #1) 10402 496 4.0 3.3780074 The Clan of the Cave Bear (Earth's Children, #1) 9804 496 5.0 3.689182 The Clan of the Cave Bear (Earth's Children, #1) 22602 496 3.0 3.3498814 The Clan of the Cave Bear (Earth's Children, #1) 41747 496 3.0 3.4196923 The Clan of the Cave Bear (Earth's Children, #1) 32576 496 5.0 3.6654968 The Clan of the Cave Bear (Eart

In [60]:
rmse = evaluator.evaluate(predictions)
print("Root-mean-square error = " + str(rmse))

Root-mean-square error = 0.9136908117752192

**Root-mean-square error = 0.913, On average the mean error is .9 that is the difference between the original rating and the predicted rating. Now, let's predict 10 books and ratings for each user.**

In [62]:
# Generate n recommendations for all users
ALS_recommendations = model.recommendForAllUsers(numItems = 10) # n - 10

In [63]:
ALS_recommendations.show(n = 10)

+-------+--------------------+
user_id| recommendations|
+-------+--------------------+
 148|[[4778, 4.4863176...|
 463|[[2627, 5.0448065...|
 471|[[1338, 4.017694]...|
 496|[[1338, 5.923006]...|
 833|[[9942, 5.1651993...|
 1088|[[9531, 4.937608]...|
 1342|[[7789, 5.308443]...|
 1580|[[4868, 5.307861]...|
 1591|[[6413, 4.853365]...|
 1645|[[9252, 4.545126]...|
+-------+--------------------+
only showing top 10 rows

In [64]:
# Temporary table
ALS_recommendations.registerTempTable("ALS_recs_temp")

In [65]:
clean_recs = spark.sql("""SELECT user_id,
                            movieIds_and_ratings.book_id AS book_id,
                            movieIds_and_ratings.rating AS prediction
                        FROM ALS_recs_temp
                        LATERAL VIEW explode(recommendations) exploded_table
                            AS movieIds_and_ratings""")
clean_recs.show()

+-------+-------+----------+
user_id|book_id|prediction|
+-------+-------+----------+
 148| 4778| 4.4863176|
 148| 9842| 4.447763|
 148| 8233| 4.43292|
 148| 6590| 4.427131|
 148| 6435| 4.4253044|
 148| 3520| 4.3928666|
 148| 1451| 4.3633265|
 148| 8946| 4.3481536|
 148| 2865| 4.333218|
 148| 6226| 4.327692|
 463| 2627| 5.0448065|
 463| 641| 4.920231|
 463| 4344| 4.757169|
 463| 347| 4.736475|
 463| 3331| 4.7246222|
 463| 6784| 4.7192845|
 463| 6032| 4.706223|
 463| 332| 4.6923423|
 463| 4201| 4.677836|
 463| 8729| 4.6729965|
+-------+-------+----------+
only showing top 20 rows

In [66]:
# Recommendations for unread books
(clean_recs.join(ratings, ["user_id", "book_id"], "left")
    .filter(ratings.rating.isNull()).show())

+-------+-------+----------+------+
user_id|book_id|prediction|rating|
+-------+-------+----------+------+
 22| 7547| 4.304169| null|
 27| 1198| 4.91851| null|
 32| 8362| 4.9440556| null|
 63| 3753| 4.6486487| null|
 85| 2652| 3.3583734| null|
 102| 9790| 2.9214456| null|
 103| 6590| 4.182767| null|
 150| 2940| 3.9354944| null|
 163| 8435| 4.819208| null|
 178| 6876| 4.6961846| null|
 200| 7550| 4.1670904| null|
 235| 7593| 3.6903093| null|
 244| 5335| 3.9928238| null|
 268| 6963| 5.2813425| null|
 277| 9570| 4.4697027| null|
 284| 7401| 4.905814| null|
 297| 6312| 4.4401374| null|
 298| 296| 4.9835024| null|
 320| 4441| 5.3855753| null|
 325| 6089| 5.4266243| null|
+-------+-------+----------+------+
only showing top 20 rows

In [67]:
new_books = (clean_recs.join(ratings, ["user_id", "book_id"], "left")
    .filter(ratings.rating.isNull()))

In [68]:
print(new_books.count())

522033

In [69]:
new_books.show(5)

+-------+-------+----------+------+
user_id|book_id|prediction|rating|
+-------+-------+----------+------+
 22| 7547| 4.304169| null|
 27| 1198| 4.91851| null|
 32| 8362| 4.9440556| null|
 63| 3753| 4.6486487| null|
 85| 2652| 3.3583734| null|
+-------+-------+----------+------+
only showing top 5 rows

In [70]:
to_read.show(5)

+-------+-------+
user_id|book_id|
+-------+-------+
 1| 112|
 1| 235|
 1| 533|
 1| 1198|
 1| 1874|
+-------+-------+
only showing top 5 rows

In [71]:
# Create Temporary Tables
new_books.createOrReplaceTempView("new_books")

In [72]:
print(new_books.count())

522033

In [73]:
print(to_read.count())

912705

In [74]:
# Create Temporary Tables
to_read.createOrReplaceTempView("to_read")

In [75]:
recommendations = new_books.join(to_read,
                              on = ["user_id", "book_id"], 
                              how = "inner")
print(recommendations.show())

+-------+-------+----------+------+
user_id|book_id|prediction|rating|
+-------+-------+----------+------+
 1266| 2840| 4.29498| null|
 22372| 7529| 4.0330534| null|
 38364| 3294| 4.050369| null|
 45435| 3849| 4.86768| null|
 11321| 8187| 4.010494| null|
 26207| 1618| 4.3792477| null|
 42077| 2700| 4.8624945| null|
 1394| 3885| 4.5378723| null|
 21924| 6425| 5.3291073| null|
 31839| 862| 5.177144| null|
 40970| 5607| 4.0587697| null|
 43735| 192| 5.316787| null|
 6713| 3124| 4.092127| null|
 44066| 1198| 4.1666613| null|
 16329| 2334| 4.752753| null|
 22982| 3243| 4.7963924| null|
 37489| 5881| 5.3499827| null|
 42874| 6384| 4.566016| null|
 19783| 6963| 4.308795| null|
 33003| 7670| 4.9878936| null|
+-------+-------+----------+------+
only showing top 20 rows

None

In [76]:
print(recommendations.count())

857

In [77]:
(recommendations
     .withColumn('pred_trunc', recommendations.prediction.substr(1,1))
     .groupby('pred_trunc')
     .count()
     .sort('pred_trunc')
    .show())

+----------+-----+
pred_trunc|count|
+----------+-----+
 0| 2|
 1| 5|
 2| 14|
 3| 113|
 4| 459|
 5| 260|
 6| 4|
+----------+-----+

In [78]:
# Create Temporary Tables
recommendations.createOrReplaceTempView("recommendations")

In [79]:
import matplotlib.pyplot as plt
plt.figure(figsize=(12,6))
plt.title('Most of the predicted ratings are above 3.8', fontsize = 12)
plt.suptitle('Distribution of predictedd ratings for the to_do lists', fontsize = 18)
rec = recommendations.toPandas()
rec['prediction'].hist()
display()

**Most of the books in the wishlist are predicted as 4.5 or more which is expected.**